### PR TITLE
fix: register indices as buffer in dense

### LINKED ISSUE #22: FixedDenseConnections.indices not saved in state_dict()
## Root Cause

`FixedDenseConnections` stores its input wiring tensor as a plain Python attribute:

```python
# connections.py, line 105
self.indices = self._init_connections()
```

PyTorch's `state_dict()` only captures `nn.Parameter` objects and registered buffers. Plain attributes are silently ignored, meaning `indices` is **never included** when saving with:

```python
torch.save(model.state_dict(), path)
```

On reload, a new model instance is created and new random connections are generated. `load_state_dict()` restores the weights but leaves the connections randomized. The loaded model now has weights optimized for one wiring applied to a completely different wiring, producing essentially random outputs.

---

## Symptoms

- Model reports correct accuracy during training (connections live in memory throughout).
- After save + reload, accuracy collapses to near-random (in our case: **88% → ~50%**, win rate **0%** in Minesweeper).
- Bug is **silent**.

---

## Reproduction

```python
model = LogicDense(in_dim=300, out_dim=2048, lut_rank=4, connections="fixed")
x = torch.randn(8, 300)
out_before = model(x)

torch.save(model.state_dict(), "model.pt")

model2 = LogicDense(in_dim=300, out_dim=2048, lut_rank=4, connections="fixed")
model2.load_state_dict(torch.load("model.pt"))

# Different output — wrong connections after reload
assert torch.allclose(out_before, model2(x))  # FAILS
```

---

## Fix

```python
# connections.py
self.register_buffer('indices', self._init_connections())
```

---

## Why It Wasn't Caught

The bug is not visible in `experiments/train.py` because evaluation is always done **in-process** without reloading. The only reload in the codebase (`generate_test_vectors.py`) correctly uses `torch.load(model)` (full object pickle), which **does** preserve plain attributes, masking the issue entirely.

### PR DESCRIPTION
**Pull request by Arash-ethz-1**
## Problem
`FixedDenseConnections` and `LearnableDenseConnections` store their wiring tensor as a plain Python attribute (`self.indices = ...`). Plain attributes are silently ignored by `state_dict()`, so connections are never saved. On reload, fresh random connections are generated and the loaded weights are applied to the wrong wiring, producing garbage outputs.

## Fix
Changed both assignments to `self.register_buffer('indices', ...)`. In the `LearnableDenseConnections` `num_candidates=-1` branch, `.contiguous()` is added because `.expand()` returns a non-contiguous view.

## Not fixed
`FixedConvConnections` has the same bug but was not touched here, its `_init_connections` returns a list of tensors with different shapes, which cannot be registered as a single buffer. Requires a separate refactor.

Closes #22

### CODE REVIEW

**Comment by ligerlac** [comment]
@Arash-ethz-1 thanks for this PR. Can you also add a test for the round trip (create model, safe to disk, reload, assert outputs are identical)?

**Comment by Arash-ethz-1** [comment]
@ligerlac sure, Ill do it tonight.

**Comment by ligerlac** [comment]
Now also fixes the issue for `FixedConvConnections`.

**Review by ligerlac — APPROVED** [review]

**Comment by Arash-ethz-1** [comment]
@ligerlac Hi, sorry I completely forgot to add the test. thanks for doing it. I also found another issue and will add it as soon as I have time.

**Comment by ligerlac** [comment]
Hi @Arash-ethz-1 no problem. Looking forward to your next contribution.
